core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -3955,11 +3960,473 @@ class KafkaApis(val requestChannel: RequestChannel,
+              subMap += tp.partition -> partitionData.errorCode
+            case None =>
+              val partitionAcknowledgementsMap : mutable.Map[Int, Short] = mutable.Map()
+              partitionAcknowledgementsMap += tp.partition -> partitionData.errorCode
+              topicPartitionAcknowledgements += tp.topicId -> partitionAcknowledgementsMap
+          }
+        }
+      }
+
+      shareFetchResponse.data.responses.forEach(topic => {
+        val topicId = topic.topicId
+        topicPartitionAcknowledgements.get(topicId) match {
+          case Some(subMap) =>
+            topic.partitions.forEach { partition =>
+              subMap.get(partition.partitionIndex) match {
+                case Some(value) =>
+                  val ackErrorCode = if(acknowledgeError.code != Errors.NONE.code) acknowledgeError.code else value
+                  partition.setAcknowledgeErrorCode(ackErrorCode)
+                  // Delete the element.
+                  subMap.remove(partition.partitionIndex)
+                case None =>
+              }
+            }
+            // Add the remaining acknowledgements.
+            subMap.foreach { case (partitionIndex, value) =>
+              val ackErrorCode = if(acknowledgeError.code != Errors.NONE.code) acknowledgeError.code else value
+              val fetchPartitionData = new ShareFetchResponseData.PartitionData()
+                .setPartitionIndex(partitionIndex)
+                .setErrorCode(Errors.NONE.code)
+                .setAcknowledgeErrorCode(ackErrorCode)
+              topic.partitions.add(fetchPartitionData)
+            }
+            topicPartitionAcknowledgements.remove(topicId)
+          case None =>
+        }
+      })
+      // Add the remaining acknowledgements.
+      topicPartitionAcknowledgements.foreach{ case(topicId, subMap) =>
+        val topicData = new ShareFetchResponseData.ShareFetchableTopicResponse()
+          .setTopicId(topicId)
+        subMap.foreach { case (partitionIndex, value) =>
+          val ackErrorCode = if(acknowledgeError.code != Errors.NONE.code) acknowledgeError.code else value
+          val fetchPartitionData = new ShareFetchResponseData.PartitionData()
+            .setPartitionIndex(partitionIndex)
+            .setErrorCode(Errors.NONE.code)
+            .setAcknowledgeErrorCode(ackErrorCode)
+          topicData.partitions.add(fetchPartitionData)
+        }
+        shareFetchResponse.data.responses.add(topicData)
+      }
+
+      if (shareSessionEpoch == ShareFetchMetadata.FINAL_EPOCH) {
+        sharePartitionManager.releaseAcquiredRecords(groupId, memberId).
+          whenComplete((releaseAcquiredRecordsData, throwable) => {
+            if (throwable != null) {
+              debug(s"Release acquired records on share session close with correlation from client ${request.header.clientId}  " +
+                s"failed with error ${throwable.getMessage}")
+              requestHelper.handleError(request, throwable)
```

Comment:
Isn't the response is already returned and this is async call, what does `requestHelper.handleError(request, throwable)` does then?

Reply:
Yep, this doesn't make sense. If there is an error here, the broker will send out 2 responses for that. I have changed this to the following logic - 
1) if throwable is not null, it will simply throw the throwable.
2) I have surrounded the call of `combineShareFetchAndShareAcknowledgeResponses` with a try catch. If error is not thrown then `requestChannel.sendResponse` with appropriate arguments. If error is thrown, then `requestHelper.handleError(request, throwable)` with appropriate arguments.

Also change the log from debug to error here.